bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCrockpotHandler.java

```diff
@@ -0,0 +1,323 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.wemo.internal.handler;
+
+import static org.openhab.binding.wemo.internal.WemoBindingConstants.*;
+
+import java.math.BigDecimal;
+import java.net.URL;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.io.transport.upnp.UpnpIOParticipant;
+import org.eclipse.smarthome.io.transport.upnp.UpnpIOService;
+import org.openhab.binding.wemo.internal.http.WemoHttpCall;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link WemoCrockpotHandler} is responsible for handling commands, which are
+ * sent to one of the channels and to update their states.
+ *
+ * @author Hans-Jörg Merk - Initial contribution;
+ */
+
+public class WemoCrockpotHandler extends AbstractWemoHandler implements UpnpIOParticipant {
+
+    private final Logger logger = LoggerFactory.getLogger(WemoCrockpotHandler.class);
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_CROCKPOT);
+
+    private Map<String, Boolean> subscriptionState = new HashMap<String, Boolean>();
+
+    private final Map<String, String> stateMap = Collections.synchronizedMap(new HashMap<String, String>());
+
+    private UpnpIOService service;
+
+    /**
+     * The default refresh interval in Seconds.
+     */
+    private final int DEFAULT_REFRESH_INTERVAL = 120;
+
+    private ScheduledFuture<?> refreshJob;
+
+    private final Runnable refreshRunnable = new Runnable() {
+
+        @Override
+        public void run() {
+            try {
+                if (!isUpnpDeviceRegistered()) {
+                    logger.debug("WeMo UPnP device {} not yet registered", getUDN());
+                }
+
+                updateWemoState();
+                onSubscription();
+            } catch (Exception e) {
```

Comment:
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

Reply:
Refactored